monai/application/engine/multi_gpu_supervised_trainer.py

```diff
@@ -0,0 +1,95 @@
+# Copyright 2020 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import torch
+
+from ignite.engine import create_supervised_trainer, create_supervised_evaluator, _prepare_batch
+
+import monai
+
+
+def _default_transform(x, y, y_pred, loss):
+    return loss.item()
+
+def _default_eval_transform(x, y, y_pred): 
+    return y_pred, y
+
+@monai.utils.export("monai.application.engine")
+def create_multigpu_supervised_trainer(net, optimizer, loss_fn, devices=None, non_blocking=False, 
+                                       prepare_batch=_prepare_batch, output_transform=_default_transform):
+    """
+    ***Derived from `create_supervised_trainer` in Ignite.
+
+    Factory function for creating a trainer for supervised models.
+    Args:
+        net (`torch.nn.Module`): the network to train.
+        optimizer (`torch.optim.Optimizer`): the optimizer to use.
+        loss_fn (torch.nn loss function): the loss function to use.
+        devices (list, optional): device(s) type specification (default: None).
+            Applies to both model and batches. None is all devices used, empty list is CPU only.
+        non_blocking (bool, optional): if True and this copy is between CPU and GPU, the copy may occur asynchronously
+            with respect to the host. For other cases, this argument has no effect.
+        prepare_batch (callable, optional): function that receives `batch`, `device`, `non_blocking` and outputs
+            tuple of tensors `(batch_x, batch_y)`.
+        output_transform (callable, optional): function that receives 'x', 'y', 'y_pred', 'loss' and returns value
+            to be assigned to engine's state.output after each iteration. Default is returning `loss.item()`.
+    Note: `engine.state.output` for this engine is defind by `output_transform` parameter and is the loss
+        of the processed batch by default.
+    Returns:
+        Engine: a trainer engine with supervised update function.
+    """
+
+    if devices is None:
+        devices = [torch.device('cuda:%i' % d) for d in range(torch.cuda.device_count())]
+    elif len(devices) == 0:
```

Comment:
You should change this to "if" instead of "elif".

Consider the case that a machine don't have GPU, and devices was None in the call. Your statement in L52 will return an empty array. This will cause problem at L59 for "devices[0]"

Reply:
Hi @anfeng  and @ericspod ,

As this file is "**multi_gpu**_supervised_trainer.py", I think we should only support **Multi-GPU** here.
Just assert something like:
```py
if devices is None:
    devices = [torch.device('cuda:%i' % d) for d in range(torch.cuda.device_count())]
assert len(devices) > 1, 'must have more 1 GPU devices.'
... ...
```
What do you think?
Thanks.

Reply:
In the case where there are not GPUs but they were requested, the list will be empty and the failure will occur on line 59. Leaving an error to propagate elsewhere is sort of the Pythonic way but instead we should raise an error. Assert isn't appropriate here as we're checking input essentially rather than an internal property our own logic should enforce if correct. I've committed changes to reflect this. I've also added code to the multi-GPU unit test to suppress the warning about GPU memory imbalance, I'd say unit tests should be silent unless they fail but maybe the warning should still be allowed or at least logged.

@Nic-Ma It's harmless to support only 1 GPU, if people have code parametrized by the number of GPUs to use they'd have to choose which function to call based on a check if that was 1 or not, it's easier if this isn't restricted in that way.

Reply:
Hi @ericspod ,

Thanks for your explanation.
If you want to suppport both CPU and GPU/GPUs logics here, why not rename the function to some general-purpose trainer? Maybe:
```py
def create_supervised_trainer(devices):
    if devices is None:
        devices = [torch.device('cuda:%i' % d) for d in range(torch.cuda.device_count())]
    if len(devices) == 0:
        devices = [torch.device("cpu")]
    ... ...
```

And about your new util function **get_devices_spec()**, I don't think it's a good idea to use **empty list** as parameter for CPU device, it's confusing.Maybe add another flag for CPU/GPU directly?
Thanks.

Reply:
@Nic-Ma What do you suggest for parameter instead? 

We can use strings like "multi-gpu", "gpu", "cpu" but if the user passes "cuda:0" what do we do?. We can enforce only using our defined names but that seems too restrictive.

Reply:
Hi @madil90 and @ericspod ,

Thanks for your comments, I prefer to this strategy:
```py
def create_supervised_trainer(devices=None):
    if devices is None:
        devices = [torch.device('cuda:%i' % d) for d in range(torch.cuda.device_count())]
        if len(devices) == 0:
            devices = [torch.device("cpu")]
    else:
        # use devices parameter directly.
    ... ...
# use cases:
trainer = create_supervised_trainer()  # automatically select devices
trainer = create_supervised_trainer(devices=[torch.device("cpu")])
trainer = create_supervised_trainer(devices=[torch.device("cuda:0")])
trainer = create_supervised_trainer(devices=[torch.device("cuda:0"), torch.device("cuda:1")])
```
1. If user pass something through "devices" parameter, let's use it directly.
    We can add some sanity check in later version.
2. If no devices provided, we try to use all GPUs first, if no GPU found, use CPU instead.

What do you think?
Thanks.

Reply:
@Nic-Ma A little verbose but seems fine. However, the last one becomes redundant. PyTorch will create a parallel context on all GPUs. The responsibility is on user to select GPUs through CUDA_VISIBLE_DEVICES.

Reply:
Hi @madil90 , Yes, thanks for your reminder.
I just want to share some alternative proposals.
For the MVP version, I am OK to use @ericspod 's method for CPU device(_empty arrary_).

Hi @ericspod , I think you can review Adil's PR for your branch and make the final solution.
Thanks.

Reply:
For the name, I used a different name to not clash with the one from Ignite, it should be obvious which is being used when looking at source code.

Choosing to use CPU computation silently when no GPU is present is going to cause people to use CPU when they didn't expect it. There should be a loud and clear error when something requested isn't possible.